blocks/api/factory.js

```diff
@@ -62,6 +67,44 @@ export function createBlock( name, blockAttributes = {} ) {
 }
 
 /**
+ * Returns an array of possible block transformations that could happen on the set of blocks received as argument.
+ *
+ * @param  {Array}  blocks Blocks array
+ * @return {Array}         Array of possible block transformations
+ */
+export function getPossibleBlockTransformations( blocks ) {
+	const sourceBlock = first( blocks );
+	if ( ! blocks || ! sourceBlock ) {
+		return [];
+	}
+	const isMultiBlock = blocks.length > 1;
+	const sourceBlockName = sourceBlock.name;
+
+	if ( isMultiBlock && ! every( blocks, { name: sourceBlockName } ) ) {
+		return [];
+	}
+
+	const blocksToBeTransformedFrom = filter( getBlockTypes(), type =>
+		find(
+			get( type, 'transforms.from', [] ),
+			t => t.type === 'block' && t.blocks.indexOf( sourceBlockName ) !== -1 &&
+			( ! isMultiBlock || t.isMultiBlock )
+		)
+	).map( type => type.name );
+
+	const blockType = getBlockType( sourceBlockName );
+	const transformsTo = get( blockType, 'transforms.to', [] );
+	const blocksToBeTransformedTo = flatMap(
+		isMultiBlock ? filter( transformsTo, transformation => transformation.isMultiBlock ) : transformsTo,
+		transformation => transformation.blocks
+	);
+
+	return uniq( compact(
```

Comment:
There's some benefit here to clarify intent, at least with the unique (we expected duplicates, we expect some falsey types), but it seems this could be simplified to a reduce where we check for those within the iterator of the reduce:

```js
return reduce( [
	...blocksToBeTransformedFrom,
	...blocksToBeTransformedTo,
], ( result, name ) => {
	const transformBlockType = getBlockType( name );
	if ( transformBlockType && ! includes( result, transformBlockType ) ) {
		result.push( transformBlockType );
	}

	return result;
}, [] );
```

Reply:
>There's some benefit here to clarify intent, at least with the unique (we expected duplicates, we expect some falsey types), but it seems this could be simplified to a reduce where we check for those within the iterator of the reduce:

Maybe worth pointing out that this might be more of a personal preference thing, and a desire to reduce the number of iterating operations, and that others might find your original approach to be clearer. I think the spread syntax could be a nice touch at least to clarify the joined set.